atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ReloadingCloseableContainer.java

```diff
@@ -0,0 +1,222 @@
+            T resource = resourceFactory.apply(factoryParameter);
+            // This lock does not need to be fair, since the isClosed lock is fair (preventing #close from being
+            // starved) and since subscribe happens after a map (and thus a new resource is created before attempting
+            // to close the old), there will not be any more readers of a resource about to be closed.
+            ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+            return new Pair<>(readWriteLock, resource);
+        }));
+
+        // The old resource is closed in a subscriber (rather than map) so that attempting to close the old resource
+        // happens after the refreshable was updated to use a new resource. This avoids a situation where
+        // a #runWithResource call using a resource that is closed, violating guarantee 1 (between the close lock
+        // being released and map completing the update).
+
+        // Furthermore, we don't want updates to have lock contention with a #runWithResource call:
+
+        // Consider the case of a long-running #runWithResource call. The current resource lock is held.
+        // If we attempt to close the resource in the map, then this must occur before refreshableResourceWithLock is
+        // updated (assuming we don't run close asynchronously). As close holds the write lock for the
+        // resource, subsequent #runWithResource calls would now block on acquiring the lock. This would be doubly
+        // damaging if the resource lock was fair, as now #runWithResource calls are blocked until all calls prior to
+        // closing the resource have finished.
+
+        // To avoid this, we must not block an update with close. This can either be done with another thread, or
+        // simply using a subscriber on the updated resource.
+        refreshableSubscriptionDisposable = refreshableResourceWithLock.subscribe(resourceWithLock -> {
+            Optional<Pair<ReadWriteLock, T>> maybeResourceWithLockToClose =
+                    currentResourceWithLock.getAndSet(Optional.of(resourceWithLock));
+            maybeResourceWithLockToClose.ifPresent(this::closeResourceAndLock);
+        });
+    }
+
+    private void closeResourceAndLock(Pair<ReadWriteLock, T> resourceWithLockToClose) {
+        Lock writeLock = resourceWithLockToClose.getLhSide().writeLock();
+        runWithLock(writeLock, () -> {
+            try {
+                T resourceToClose = resourceWithLockToClose.getRhSide();
+                resourceToClose.close();
+            } catch (IOException e) {
+                log.warn("Failed to close resource. This may result in a resource leak", e);
+            }
+        });
+    }
+
+    /***
+     * Closes the container and the latest created resource.
+     * After this method has returned, no further resources will be created in this container, and all subsequent
```

Comment:
is this actually true? Does this actually stop the _map_ function from running from here on out? I didn't see that happening (I can believe that the subscriber stops, but not sure about the map), and so it is possible that the refreshable actually continues to refresh, and that new values continue to be computed and leaked.

Reply:
This test fails (when run in your test class):
```java
        ReloadingCloseableContainer<Closeable> container = getContainer();
        container.close();
        settableRefreshable.update(UPDATED_VALUE);
        verify(mockCloseable).close();
        verify(mockCloseableExtra).close();
```

You might want to either adjust your javadoc, or (more likely) think about doing something to guarantee that we stop creating and leaking resources!

Reply:
The map function is wrapped in a runIfNotClosed. Your test should fail, because the second resource should never be created in the first place - and verifying that a non-existent resource is closed is a contradiction. `ReloadingCloseableContainer#noFurtherResourceCreatedAfterClose` is what, I believe, the test should be.

Reply:
The map function is wrapped in a runIfNotClosed. Your test should fail, because the second resource should never be created in the first place - and verifying that a non-existent resource is closed is a contradiction. `ReloadingCloseableContainer#noFurtherResourceCreatedAfterClose` is what, I believe, the test should be.

Reply:
I think you're right, although I'm not 100% sure whether the current test actually fully covers this behaviour. Still, I hadn't reviewed all of the tests by this point, so a fair flag that I should have seen this. I think what I'd want to confirm is that, when the refreshable changes after being created, the factory no longer does anything - the test you linked shows that at the end, we've applied once to `any()` - let's make this:
1. not any, but the specific value we expect
2. confirm that that is called _before_ the close, and then that there are no more interactions after the close.